---
title: "generalCaseRNASeqAnalysis"
author: "Jared Hendrickson"
date: "2023-02-27"
output: html_document
params:
  study: "UCI_3xTg-AD"
  control: "B6129F3"
  treatment: "3xTg-AD"
  synapseusername: "***"
  synapsepassword: "***"
  geneCountsTableSynID: "syn51200765"
  MODELADHarmonizationSynID: "syn51040095"
---



```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

```

### These are the minimal required packages

```{r clear workspace and load r packages, include=FALSE}

library(tidyverse)
library(DESeq2)
library(glue)
library(synapser)
library(reticulate)

# log into Synapse

synLogin(params$synapseusername, params$synapsepassword)

```

### Use external python file to pull metadata

```{r save metadata for study}

source_python("captureMetadataReticulate.py")

saveMetadata(params$synapseusername, params$synapsepassword, params$study)

```
### Samplesheet preparation in Nextflow

```{r samplesheet preparation in Nextflow}

studiesRepresented <- c("Jax.IU.Pitt_APP.PS1", "MODEL-AD_JAX_GWAS_Gene_Survey", "UCI_5XFAD", "Jax.IU.Pitt_APOE4.Trem2.R47H", "UCI_hAbeta_KI", "Jax.IU.Pitt_PrimaryScreen", "Jax.IU.Pitt_StrainValidation", "Jax.IU.Pitt_hTau_Trem2", "Jax.IU.Pitt_5XFAD", "UCI_3xTg-AD", "UCI_PrimaryScreen")

y <- paste0("'", studiesRepresented, "'")
x <- paste(y, collapse = ",")

input <- synTableQuery(glue("select * from syn11346063 WHERE (\"study\" HAS ({x})) AND (\"assay\" HAS ('rnaSeq')) AND (\"fileFormat\" = 'fastq')"))

data <- read.csv(input$filepath)

# sanity checks

data$name %>%
  length()

data$name %>%
  unique() %>%
  length()

data$specimenID %>%
  unique() %>%
  length()

data$individualID %>%
  unique() %>%
  length()

# some wrangling/testing

test <- data %>%
  filter(!str_detect(name, "R")) %>%
  select(name) %>%
  mutate(name = name %>% str_replace_all("_1", "_R1")) %>%
  mutate(name = name %>% str_replace_all("_2", "_R2"))

# create map between original names and RNAseq pipeine acceptable names

neededData <- data %>%
  select(sample = specimenID, name, id) %>%
  mutate(OriginalSample = sample) %>%
  mutate(sample = sample %>% str_remove_all(",") %>% str_remove_all("\\(") %>% str_remove_all("\\)"))

# raw data pulled from Synapse table to be wrangled for samplesheet
write.csv(neededData, "neededData.csv", row.names = FALSE)

neededData <- neededData %>%
  select(sample, name, id)

fastq1 <- neededData %>%
  filter(str_detect(name, "R1|_1\\.")) %>%
  mutate(fastq1 = paste0("syn://", id)) %>%
  select(sample, fastq1)

fastq2 <- neededData %>%
  filter(str_detect(name, "R2|_2\\.")) %>%
  mutate(fastq2 = paste0("syn://", id)) %>%
  select(sample, fastq2)

sampleSheet <- left_join(fastq1, fastq2, by = "sample") %>%
  mutate(strandedness = "auto") %>%
  dplyr::rename(fastq_1 = fastq1, fastq_2 = fastq2)

write.csv(sampleSheet, "studyManifestSampleSheet.csv", row.names = FALSE)

```
### Load gene counts file and raw data pulled from studies represented Synapse table

```{r load RSEM counts from nf-core rnaseq pipeline}

fileEntity <- synGet(params$geneCountsTableSynID)
counts <- read.csv(fileEntity$path, sep = "\t")

originalNameMap <- read.csv("neededData.csv")

```
### Nextflow pipeline renames some specimen ids
### Wrangling functions to map back

```{r functions to transform specimenIDs in counts matrix to inputted specimenIDs}

trimX <- function(specimenID) {
  if(substr(specimenID, 1, 1) == "X") {
    return(substr(specimenID, 2, nchar(specimenID)))
  }
  else {
    return(specimenID)
  }
}

replacePeriod <- function(specimenID) {
  return(str_replace(specimenID, "\\.", "-"))
}

```

### Wrangle Nextflow gene counts specimen ids
### Do some sanity checks

```{r checks}

countMap <- data.frame(Nextflow_specimenIDs = colnames(counts), inputSpecimenIDS = NA)

countMap <- countMap %>%
  filter(!(Nextflow_specimenIDs %in% c("transcript_id", "gene_id")))

for(i in 1:nrow(countMap)) {
  countMap[i, "inputSpecimenIDS"] <- countMap[i, "Nextflow_specimenIDs"] %>% 
    trimX() %>% 
    replacePeriod()
}

# sanity checks

intersect(countMap$inputSpecimenIDS, originalNameMap$sample) %>%
  length()

setdiff(countMap$inputSpecimenIDS, originalNameMap$sample) %>%
  length()

setdiff(originalNameMap$sample, countMap$inputSpecimenIDS) %>%
  length()

# future development: sanity check fastq specimen ids and biospecimen specimen ids

```
### Load study metadata

```{r load metadata}

metadata <- read.csv(paste0(params$study, "_", "joinedMetadata.csv"), stringsAsFactors = FALSE)

```

### More Tests/intersect length should be number of RNAseq samples in study

```{r general intersect tests}

# should match number of rows in assay/rnaseq metadata
# p adjust (check)

intersect(metadata$specimenID, countMap$inputSpecimenIDS)

intersect(metadata$specimenID, countMap$inputSpecimenIDS) %>% length()

```
### Only include metadata that we have rnaseq data for

```{r only include metadata that we have rnaseq data for}

metadataRNASeqOnly <- metadata %>%
  filter(specimenID %in% countMap$inputSpecimenIDS)

```


### Reduce counts matrix to study samples only

```{r reduce counts matrix to study samples only}

filtCountMap <- countMap %>%
  filter(inputSpecimenIDS %in% metadataRNASeqOnly$specimenID)

counts_reduced_to_study <- counts %>%
  select(gene_id, filtCountMap$Nextflow_specimenIDs)

```

### Prepare gene counts table for Deseq2

```{r prepare counts}

samples <- colnames(counts_reduced_to_study)[3:length(counts_reduced_to_study)]

rownames(counts) <- counts$gene_id

counts <- counts %>%
  select(-gene_id)

```

### Prepare Metadata

```{r prepare metadata}

rownames(metadataRNASeqOnly) <- filtCountMap$Nextflow_specimenIDs
metadataRNASeqOnly$Nextflow_specimenIDs <- rownames(metadataRNASeqOnly)

# sex, age, tissue

```

### set up stratifications and test set

```{r set up stratifications and test set}

metadataRNASeqOnly_renamed <- metadataRNASeqOnly %>%
  dplyr::rename(age = ageDeath)

metadataRNASeqOnly_renamed$genotype <- metadataRNASeqOnly_renamed$individualCommonGenotype %>%
  as.factor()

stratifications <- split(metadataRNASeqOnly_renamed, list(metadataRNASeqOnly_renamed$sex, metadataRNASeqOnly_renamed$age, metadataRNASeqOnly_renamed$tissue))

# for(i in 1:length(stratifications)) {
#   print(stratifications[[i]])
# }

metadataStrata <- stratifications[[1]]
countsStrata <- counts %>%
  select(metadataStrata$Nextflow_specimenIDs)

write.csv(metadataStrata, "metadataStrataEX.csv", row.names = FALSE)
write.csv(countsStrata, "countsStrataEx.csv", row.names = FALSE)

```

### Deseq2 Design

```{r runDeseq2 design}

runDeseq2 <- function(counts, metadataStrata) {
  
  countsStrata <- counts %>%
  select(metadataStrata$Nextflow_specimenIDs)
  
  dds <- DESeqDataSetFromMatrix(countData = round(countsStrata),
                                colData = metadataStrata,
                                design= ~ genotype)
  
  design(dds) <- formula(~ genotype)

  dds <- DESeq(dds)

  return(dds)
}

```

### Output dataframe

```{r dataframe output}

createDataFrameOutput <- function(dds, metadataStrata, control, treatment) {
  
  metadataRow <- metadataStrata %>%
    dplyr::distinct(sex, age, tissue) %>%
    mutate(strain = treatment)
  
  res <- results(dds, contrast = c("genotype", treatment, control))
  
  res_df <- as.data.frame(res) %>%
    select(log2FoldChange, pvalue, padj) %>%
    filter(!is.na(log2FoldChange), !is.na(pvalue), !is.na(padj)) %>%
    mutate(geneID = rownames(.))
  
  # may need to change filter
  
  rownames(res_df) <- NULL
  
  metadataRow <- metadataRow %>%
    dplyr::distinct(sex, age, tissue) %>%
    mutate(strain = treatment)
  
  metadataRowExpanded <- do.call("rbind", replicate(nrow(res_df), metadataRow, simplify = FALSE))
  
  rownames(metadataRowExpanded) <- NULL
  
  final <- cbind(metadataRowExpanded, res_df) %>%
    dplyr::select(strain, sex, age, tissue, geneID, log2FoldChange, pvalue, padj)
  
  return(final)
}

```

### Run all stratifications

```{r run all stratifications}

dir.create(params$study)

for(x in stratifications) {
  dds <- runDeseq2(counts = counts, metadataStrata = x)
  
  df <- createDataFrameOutput(dds = dds, metadataStrata = x, control = params$control, treatment = params$treatment)
  
  write.csv(df, paste0(params$study, "/", df$sex %>% unique(), "_", df$age %>% unique(), "mon_", df$tissue %>% unique(), "_", x$specimenIdSource %>% unique(), ".csv"), row.names = FALSE)
}

```


```{r functions to combine and test against historical data}

# combine shared gene ids

mergeOutput <- function(SageOupout, UCIOutput) {
  log2FoldChangeSage <- SageOupout %>%
    dplyr::rename(log2FoldChangeSage = log2FoldChange) %>%
    select(geneID, log2FoldChangeSage)
  
  log2FoldChangeUCI <- UCIOutput %>%
    dplyr::rename(log2FoldChangeUCI = log2FoldChange) %>%
    select(geneID, log2FoldChangeUCI)
  
  combined <-
    inner_join(log2FoldChangeSage, log2FoldChangeUCI, by = "geneID")
  
  return(combined)
}

```

```{r plot Sage versus UCI center data example}

combined <- mergeOutput(SageOupout = read.csv("female_4mon_hippocampus_UCI_TMF.csv"), UCIOutput = read.csv("notSageProcessed/Female_4mon_hippocampus_UCI.csv"))

ggplot(combined, aes(x = log2FoldChangeUCI, y = log2FoldChangeSage)) +
  geom_point() +
  geom_smooth(method = "lm")

ggsave("female_4mon_hippocampus_UCIandSageProcessed.png")

cor(combined$log2FoldChangeSage, combined$log2FoldChangeUCI)

```

```{r}

res <- resid(mod <- lm(log2FoldChangeSage ~ log2FoldChangeUCI, data = combined))

res.qt <- quantile(res, probs = c(0.001,0.999))

want <- which(res >= res.qt[1] & res <= res.qt[2])

```

```{r}

outliers <- combined[-want,]
within <- combined[want,]

ggplot(within, aes(x = log2FoldChangeUCI, y = log2FoldChangeSage)) +
  geom_point() +
  geom_smooth(method = "lm")

ggsave("MainData.png")

ggplot(outliers, aes(x = log2FoldChangeUCI, y = log2FoldChangeSage)) +
  geom_point() +
  geom_smooth(method = "lm")

ggsave("outliers.png")

```
```{r}

SageOutput = read.csv("female_4mon_hippocampus_UCI_TMF.csv")
UCIOutput = read.csv("notSageProcessed/Female_4mon_hippocampus_UCI.csv")

testSage <- SageOutput %>% 
  filter(geneID %in% outliers$geneID)
testUCI <- UCIOutput %>% 
  filter(geneID %in% outliers$geneID)

```

### Set Complete Provenance

```{r Set Complete Provenance}

source_python("assignProvenance.py")

setCompleteProvenanceStudy(params$synapseusername, params$synapsepassword, params$study, params$geneCountsTableSynID, params$MODELADHarmonizationSynID)

```

